browser-test/test.endtoend.ts

```diff
@@ -0,0 +1,546 @@
+            projectId: 'bogus',
+          });
+  
+          // Mock request
+          const request = {};
+  
+          // Mock response
+          const content = 'content951530617';
+          const expectedResponse = {
+            content,
+          };
+  
+          // Mock Grpc layer
+          client._innerApiCalls.wait = mockLongRunningGrpcMethod(
+            request,
+            expectedResponse
+          );
+  
+          client
+            .wait(request)
+            .then(responses => {
+              const operation = responses[0];
+              return operation.promise();
+            })
+            .then(responses => {
+              assert.deepStrictEqual(responses[0], expectedResponse);
+              done();
+            })
+            .catch(err => {
+              done(err);
+            });
+        });
+  
+        it('invokes wait with error', done => {
+          const client = new EchoClient({
+            auth: authStub,
+            credentials: {client_email: 'bogus', private_key: 'bogus'},
+            projectId: 'bogus',
+          });
+  
+          // Mock request
+          const request = {};
+  
+          // Mock Grpc layer
+          client._innerApiCalls.wait = mockLongRunningGrpcMethod(
+            request,
+            null,
+            error
+          );
+  
+          client
+            .wait(request)
+            .then(responses => {
+              const operation = responses[0];
+              return operation.promise();
+            })
+            .then(() => {
+              assert.fail();
+            })
+            .catch(err => {
+              assert(err instanceof Error);
+              assert.strictEqual(err.code, FAKE_STATUS_CODE);
+              done();
+            });
+        });
+  
+        it('has longrunning decoder functions', () => {
+          const client = new EchoClient({
+            auth: authStub,
+            credentials: {client_email: 'bogus', private_key: 'bogus'},
+            projectId: 'bogus',
+          });
+          assert(
+            client._descriptors.longrunning.wait.responseDecoder instanceof Function
+          );
+          assert(
+            client._descriptors.longrunning.wait.metadataDecoder instanceof Function
+          );
+        });
+    });
+});
+
+describe('Run tests against gRPC server', () => {
+    const authStub = {
+        getRequestHeaders() {
+        return {Authorization: 'Bearer SOME_TOKEN'};
+        },
+      };
+    
+    const opts = {
+          auth: authStub,
+    };
+
+    const client = new EchoClient(opts);
+
+    before(async function () {
+        this.timeout(30000);
+        console.log('before block started');
+        const request = {
+            content: 'test',
+        };
+        
+        await sleep(5000);
+        const MAX_RETRIES = 20;
+        const TIMEOUT = 1000;
+
+        // for (let retryCount = 0; retryCount < MAX_RETRIES; ++retryCount) {
```

Comment:
why is this commented out? Should it be deleted? If it's meant for a later PR, please delete it. If it is to be kept, please put a comment as to why its commented out.

Reply:
I realize now that this is polling code for waiting on the server to start.

I think we should be using the callback form of `before` (or promise form, either way) and be calling `startServer`  here and giving it the `done` callback function.

WDYT?

Reply:
All the stuff that's in test.endtoend.ts is run on browser (as part of the karma framework) and most of the code in startServer() didn't work there since a lot of it can't be webpacked. Alex suggested using a child-process to start the server (in Node.js) before running any of the tests to make it work, which is why I did it this way! Do you think it makes sense as a solution?